compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -0,0 +1,537 @@
+        rewriter, op.getDestinationOperands(rewriter), offsets, sizes, true);
+    if (tiledImplementation.size() != 1) {
+      return rewriter.notifyMatchFailure(
+          op, "expected tiled implementation to return a single op");
+    }
+    tilingResult.tiledOp = tiledImplementation[0];
+
+    LLVM_DEBUG({
+      if (!tilingResult.loops.empty()) {
+        llvm::errs() << "After tiled implementation :\n";
+        tilingResult.loops.front().dump();
+        llvm::errs() << "\n";
+      }
+    });
+  }
+
+  // Update the filter.
+  filter.replaceLinalgTransformationFilter(rewriter, tilingResult.tiledOp);
+
+  if (op->getNumResults() == 0) {
+    rewriter.eraseOp(op);
+    return tilingResult;
+  }
+
+  // Rewrite all `flow.dispatch.workgroups`. First find all the stores.
+  SmallVector<IREE::Flow::DispatchTensorStoreOp> storeOps;
+  for (OpOperand &use : op->getUses()) {
+    auto storeOp = dyn_cast<IREE::Flow::DispatchTensorStoreOp>(use.getOwner());
+    if (storeOp && storeOp.getValue() == use.get()) {
+      storeOps.push_back(storeOp);
+    }
+  }
+  for (auto storeOp : storeOps) {
+    OpResult result = storeOp.getValue().cast<OpResult>();
+    SmallVector<OpFoldResult> resultOffsets, resultSizes;
+    if (failed(op.getResultTilePosition(rewriter, result.getResultNumber(),
+                                        offsets, sizes, resultOffsets,
+                                        resultSizes))) {
+      return rewriter.notifyMatchFailure(
+          op, "failed to rewrite destructive update");
+    }
+    SmallVector<OpFoldResult> resultStrides(resultOffsets.size(),
+                                            rewriter.getIndexAttr(1));
+
+    SmallVector<OpFoldResult> combinedOffsets, combinedSizes, combinedStrides;
+    if (failed(IREE::Flow::foldOffsetsSizesAndStrides(
+            rewriter, loc, storeOp.getMixedOffsets(), storeOp.getMixedSizes(),
+            storeOp.getMixedStrides(), storeOp.getDroppedDims(), resultOffsets,
+            resultSizes, resultStrides, combinedOffsets, combinedSizes,
+            combinedStrides))) {
+      return rewriter.notifyMatchFailure(
+          op, "failed to create tiled flow.dispatch.tensor.store op");
+    }
+
+    rewriter.create<IREE::Flow::DispatchTensorStoreOp>(
+        loc, tilingResult.tiledOp->getResult(result.getResultNumber()),
+        storeOp.getTarget(), storeOp.getTargetDims(), combinedOffsets,
+        combinedSizes, combinedStrides);
+  }
+  for (auto storeOp : storeOps) {
+    rewriter.eraseOp(storeOp);
+  }
+  return tilingResult;
+}
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// SwapExtractSliceWithTiledProducer
+//===----------------------------------------------------------------------===//
+
+/// Pattern to swap a `tilinginterface op` -> `tensor.extract_slice` with
+/// `tensor.extract_slice` of operands of the op -> tiled `tilinginterface op`.
+struct SwapExtractSliceWithTiledProducer
```

Comment:
IIUC, this is the pattern that actually fuses multiple tiled operations into one loop nest?

Reply:
Thats right. It swaps the slice of producer with a tiled implementation of the producer when the producer also implements the `TilingInterface`.